File: Design and Verification Training Program/Veriest/Scripting/Perl/create_module_syntax_with_num_of_bits.pl
#!/usr/bin/perl -w
use strict;
use Data::Dumper;

my $file_name = $ARGV[0];
my $module_name = $ARGV[1];
my $module_param_name = $ARGV[2];
my $module_param_value = $ARGV[3];
open (FH, '<', $file_name) or die "Couldn't open file";
my @lines = <FH>; # load lines from text file
#print "@lines \n";
my @temp;
@temp = split(/[=]+/, $lines[0]);
#print "@temp \n";
$temp[1] =~ s/\s+//g; # remove all white spaces of the temp[1] string (dict)
$temp[1] =~ s/'+//g; # remove all " ' " of the temp[1] string (dict)
#print "$temp[1] \n";
substr($temp[1], 0, 1, ''); # remove first char ({)
substr($temp[1], -1, 1, ''); # remove last char (})
#print "$temp[1] \n";

@lines = split(/[{ \s\ }]+/, $temp[1]); # split string by "{" and "}"
#print "@lines \n";
my $index;
for ($index = 1; $index <= $#lines; $index = $index + 4){
    splice @lines, $index, 0, '{'; # insert "{" as array element at the "{" split points of the string
    splice @lines, $index + 2, 0, '}';  # insert "}" as array element at the "}" split points of the string
}
#print "@lines \n";

my @new_lines;
my $item;
foreach $item (@lines){
    push @new_lines, split(/[, \s\ :]+/, $item); # split strings in lines array by "," and ":"
}
#print "@new_lines \n\n";
@new_lines = grep /\S/, @new_lines; # remove empty elements from new_lines array
#print "@new_lines \n\n";


foreach $item (@new_lines){
    if ($item =~ m/\[/){ # if "[" is in new_lines array item
	my ($indx) = grep { $new_lines[$_] ~~ $item } 0 .. $#new_lines; # find the index of the array item
	my $new_item = $new_lines[$indx].":".$new_lines[$indx + 1]; # combine "[x" and "y]" array items to one item "[x:y]"
	splice(@new_lines, $indx + 1, 1); # remove the "y]" item from the array
	splice(@new_lines, $indx, 1); # remove the "[x" item from the array
	splice(@new_lines, $indx, 0, $new_item); # insert the new item to the array in the index of "[x"
    }
}
#print "@new_lines \n";

my %signals_hash;
for ($index = 0; $index <= $#new_lines; $index++){
    if ($new_lines[$index] eq '{'){
	my %signals;
	my $i = 1;
	while ($new_lines[$index + $i] ne '}'){
	    $signals{$new_lines[$index + $i]} = $new_lines[$index + $i + 1];
	    $i = $i + 2;
	}
	$signals_hash{$new_lines[$index - 1]} = \%signals;
    }
}
#print Dumper(\%signals_hash);



my @filter_keys = 'num_of_bits';
my %num_of_bits_hash = map { $_ => $signals_hash{$_} } grep { exists $signals_hash{$_} } @filter_keys; # extract hash element with the key "num_of_bits" and store in a new hash
#print Dumper(\%num_of_bits_hash);
delete $signals_hash{$filter_keys[0]}; # remove the hash element with the key "num_of_bits" from signals_hash
#print Dumper(\%signals_hash);

my $key;
my $k;
foreach $key (keys %signals_hash){
    foreach $k (keys %{$signals_hash{$key}}){
	if ($signals_hash{$key}{$k} eq "Null"){
	    delete $signals_hash{$key}{$k}; # remove "Null" values from hash
	}
    }
}
#print Dumper(\%signals_hash);

my $file = $module_name . ".sv";
open (FILE, '>'.$file) or die "Couldn't open file";
print FILE "module $module_name\n";
print FILE "\t#(parameter $module_param_name  = $module_param_value)\n";
print FILE "\t(\n";
foreach $key (keys %signals_hash){
    foreach $k (keys %{$signals_hash{$key}}){
	my $n_of_bits = $num_of_bits_hash{$filter_keys[0]}{$k} ne "Null" ? $num_of_bits_hash{$filter_keys[0]}{$k} : "\t"; # check if the current num of bits is not Null
	if (($key eq (keys %signals_hash)[-1]) && ($k eq (keys %{$signals_hash{$key}}))[-1]){ # if the current item is the very last item in the nested hashes
	    print FILE "\t$key $n_of_bits\t$signals_hash{$key}{$k}\n";
	} else {
	    print FILE "\t$key $n_of_bits\t$signals_hash{$key}{$k},\n";
	}
    }
} 
print FILE "\t);\n";
print FILE "\n\nendmodule";
close FILE;

open (FH, $file) or die "Couldn't open file";
while (<FH>)
{
	print $_;
}
close;

File: Design and Verification Training Program/Veriest/Scripting/Python/create_module_syntax_with_num_of_bits.py
#!/usr/bin/python
import sys
import re

file_name = sys.argv[1]
module_name = sys.argv[2]
module_param_name = sys.argv[3]
module_param_value = sys.argv[4]
signals = open(file_name,'r')
lines = signals.read()
#print(lines)

# Works perfectly, but bad practice:
#context = {}
#d = exec(lines, context)

#
# Convert dictionary given as string to dictionary type:
#
# Remove "d =" from the string read from the text file:
string_dict = lines.split("=")
#print(string_dict)
string_dict = string_dict[1]
#print(string_dict)
# Remove " ", "'" and the outer "{" "}" from the string:
string_dict = string_dict.replace(" ","")
string_dict = string_dict.replace("'","")
string_dict = string_dict[1:-1]
#print(string_dict)

# Split string_dict at "{":
lbraces = string_dict.split("{")
#print(lbraces)
# Insert ":{" into lbraces where it's supposed to be at:
for i in range (1,(len(lbraces)+len(lbraces)-1),2):
    lbraces.insert(i,"{")
#print(lbraces)

# Split lbraces items at "}":
rbraces = []
for j in lbraces:
    rbraces.append(j.split("}"))
#print(rbraces)
# Insert ":{" into rbraces where it's supposed to be at:
for k in rbraces:
    for l in range (1,(len(k)),2):
        k.insert(l,'}')
#print(rbraces)

# Split items of rbraces at "," and ":":
comma_colon = []
for a in rbraces:
    for b in a:
        comma_colon.append(re.split(',|:',b))

# Flatten comma_colon list (all nested items become non-nested items in flat_list):
flat_list = []
for sublist in comma_colon:
    for item in sublist:
        flat_list.append(item)
#print(flat_list)

# Remove blank items from flat_list
while '' in flat_list:
    flat_list.remove('')
#print(flat_list)

# Combine '[x' and 'y]' list elements into '[x:y]':
for element in flat_list:
    if "[" in element:
        indx = flat_list.index(element)
        new_element = flat_list[indx]+":"+flat_list[indx+1]
        flat_list = flat_list[:indx] + flat_list[indx+2:]
        flat_list.insert(indx, new_element)
#print(flat_list)

# Create nested dict from flat_list:
d={}
for index in range (0,len(flat_list)):
    if flat_list[index] == '{':
        my_dict={}
        i=1
        #print(index)
        while flat_list[index+i] != '}':
            my_dict[flat_list[index+i]]=flat_list[index+i+1]
            i=i+2
        d[flat_list[index-1]]=my_dict
#print(d)

#d = {'input': {'1': 'haddr', '2': 'hwrite', '3': 'hsize' , '4': 'hsel', '5': 'hburst', '6': 'hprot', '7': 'htrans', '8': 'hmastlock', '8': 'hready', '9': 'hmastlock', '10': 'hwdata', '11' : 'hresetn', '12': 'hclk', '13': 'Null', '14': 'sys_clk', '15': 'sys_reset', '16': 'sys_err'},'output': {'1': 'hrdata', '2': 'hresp','3': 'hreadyout', '4': 'Null','5': 'Null', '6': 'Null','7': 'Null', '8': 'Null','9': 'Null', '10': 'Null','11': 'Null', '12': 'Null', '13': 'Null','14': 'Null', '15': 'Null', '16': 'Null'},'inout': {'1': 'datap', '2': 'datan','3': 'clkp', '4': 'clkn','5': 'Null', '6': 'Null','7': 'Null', '8': 'Null','9': 'Null', '10': 'Null','11': 'Null', '12': 'Null', '13': 'Null','14': 'Null', '15': 'Null', '16': 'Null'},'num_of_bits': {'1': '[31:0]', '2': 'Null','3': '[2:0]', '4': 'Null','5': '[2:0]', '6': '[3:0]','7': '[1:0]', '8': 'Null','9': 'Null', '10': '[31:0]','11': 'Null', '12': 'Null', '13': 'Null','14': 'Null', '15': 'Null', '16': '7:0'}}


# Split d dict into 2 dicts by keys:
num_of_bits_key = ['num_of_bits']
signals_keys = ['input','output','inout']
# these two lines will work on a newer version of python:
#[num_of_bits_data,signals_data] = map(lambda keys: {x: d[x] for x in keys}, [num_of_bits_key, signals_keys])
#filterByKey = lambda keys: {x: d[x] for x in keys}
filter_by_key = lambda keys: dict((x,d[x]) for x in keys)
num_of_bits_data = filter_by_key(num_of_bits_key)
signals_data = filter_by_key(signals_keys)

# Remove dict items with "Null" values:
delete_list = []
for k in signals_data.keys():
    for i_k in signals_data[k].keys():
        if signals_data[k][i_k] == "Null":
            delete_list.append([k,i_k])
for [x,y] in delete_list:
    del signals_data[x][y] 

#
# Create sv module template:
#
f = open(module_name+'.sv','w')
f.write("module " + module_name + "\n")
f.write("\t" + "#(parameter " + module_param_name + " = " + module_param_value + ")\n")
f.write("\t(\n")

for key in signals_data:
    for inner_key in signals_data[key]:
       # if (signals_data[key][inner_key] != "Null"):
            n_of_bits = num_of_bits_data['num_of_bits'][inner_key] if num_of_bits_data['num_of_bits'][inner_key] != "Null" else "\t"
            if ((key == signals_data.keys()[-1]) & (inner_key == signals_data[key].keys()[-1])): # if the current item is the very last item in the nested dicts
                f.write("\t" + key + " " + n_of_bits + "\t" + signals_data[key][inner_key] + "\n") 
            else:
                f.write("\t" + key + " " + n_of_bits + "\t" + signals_data[key][inner_key] + ",\n")

f.write("\t);\n")
f.write("\n\nendmodule")
f.close()
f = open(module_name+'.sv','r')
print(f.read())
